functions/nodejs-apps.bash

```diff
@@ -202,3 +202,142 @@ nodered_setup() {
+
+  serverIP="$(hostname -I)"; serverIP=${serverIP::-1} # remove trailing space
+  installSuccessText="Setup was successful. Zigbee2MQTT is now up and running.\\n\\nFor further Zigbee-settings open frontend (in 2 minutes): \\nhttp://${serverIP}:8081.\n\nDocumentation of ZigBee2MQTT:\\nhttps://www.zigbee2mqtt.io/guide/configuration"
+  updateSuccessText="Update successful. \\n\\nFor further Zigbee-settings open frontend (in 2 minutes): \\nhttp://${serverIP}:8081.\n\nDocumentation of Zigbee2MQTT:\\nhttps://www.zigbee2mqtt.io/guide/configuration"
+ 
+  if [[ $1 == "remove" ]]; then
+    if [[ -n $INTERACTIVE ]]; then
+      if ! (whiptail --title "Zigbee2MQTT Uninstall" --yes-button "Continue" --no-button "Cancel" --yesno "$uninstallText" 7 80); then echo "CANCELED"; return 0; fi
+    fi
+    echo -n "$(timestamp) [openHABian] Removing Zigbee2MQTT service... "
+    if ! cond_redirect systemctl stop zigbee2mqtt.service; then echo "FAILED (disable service)"; fi
+    if ! rm -f /etc/systemd/system/zigbee2mqtt.service; then echo "FAILED (remove service)"; fi
+    if cond_redirect systemctl -q daemon-reload; then echo "OK"; else  echo "FAILED (daemon-reload)"; return 1; fi
+
+    echo -n "$(timestamp) [openHABian] Uninstalling Zigbee2MQTT... "
+    if ! cond_redirect npm uninstall zigbee2mqtt ; then echo "FAILED (npm uninstall)"; fi
+    if ! rm -rf /var/log/zigbee2mqtt; then echo "FAILED (remove log)"; fi
+    if rm -rf "/opt/zigbee2mqtt"; then echo "OK"; else echo "FAILED"; fi
+    
+    if [[ -n "$INTERACTIVE" ]]; then
+      whiptail --title "Zigbee2MQTT removed" --msgbox "Zigbee2MQTT was removed from your system." 7 80
+    fi
+    return 0;
+  fi
+  if [[ $1 != "install" ]]; then return 1; fi
+
+  # if a config file exists do only update and exit
+  if [[ -e "/opt/zigbee2mqtt/data/configuration.yaml" ]] ; then
+    if [[ -n $INTERACTIVE ]]; then
+      if ! (whiptail --title "Zigbee2MQTT installation" --yes-button "Continue" --no-button "Cancel" --yesno "$z2mInstalledText" 14 80); then echo "CANCELED"; return 0; fi
+    fi
+
+    echo -n "$(timestamp) [openHABian] Updating Zigbee2MQTT... "
+    cd /opt/zigbee2mqtt || echo "FAILED (cd)"
+    if ! cond_redirect systemctl stop zigbee2mqtt ; then echo "FAILED (stop systemctl)"; fi
+    if ! cond_redirect sudo -u "${username:-openhabian}" cp -R data data-backup; then echo "FAILED (cp backup)"; fi
+    if ! cond_redirect sudo -u "${username:-openhabian}" git checkout HEAD -- npm-shrinkwrap.json; then echo "FAILED git"; fi
+    if ! cond_redirect sudo -u "${username:-openhabian}" git pull; then echo "FAILED git"; fi
+    if ! cond_redirect sudo -u "${username:-openhabian}" npm install zigbee2mqtt ; then echo "FAILED npm"; fi
+    if ! cond_redirect sudo -u "${username:-openhabian}" cp -R data-backup/* data; then echo "FAILED (cp backup)"; fi
+    rm -rf /opt/zigbee2mqtt/data-backup
+    cd /opt || echo "FAILED (cd)"
+    if ! cond_redirect systemctl start zigbee2mqtt; then echo "FAILED"; fi
```

Comment:
Should be returning on these lines if not successful to let the user know that something went wrong and potentially breaking a working install by continuing.

Reply:
now always returning 1. Only exception is "systemctl stop zigbee2mqtt" (line 272) because the user could have already stopped the service. I really would prefer to carry on even if the service is not running.